### PR TITLE
Ensure that concat output coherent dtypes

### LINKED ISSUE #4685: Issue with casting during concatenation of dataframe
I got an issue with the following code which suprised me:

```python
import numpy as np
import pandas as pd

import dask.dataframe as dd

pd.DataFrame(np.random.randint(10, size=100000)).to_csv('df_a.csv',
                                                        index=False)
pd.DataFrame(np.random.randn(100000)).to_csv('df_b.csv', index=False)

df_a = dd.read_csv('df_a.csv')
df_b = dd.read_csv('df_b.csv')
df = dd.concat([df_a, df_b], axis=0)

df.to_parquet('df.parquet')
df = dd.read_parquet('df.parquet')
```

The parquet parts dumped on the disk are of different dtype while I would have expect `dd.concat` to make the casting without the need to explicitly call `.compute()`.

Is it a desired behavior?

### PR DESCRIPTION
**Pull request by glemaitre**
fixes #4685

- [ ] Tests added / passed
- [ ] Passes `flake8 dask`

### CODE REVIEW

**Comment by glemaitre** [comment]
@mrocklin I see that this is fixing the issue.

However, what would be a good regression test without dumping a parquet file.
Is there an easy way to access the dtypes of the partitions after concatenation?

**Comment by mrocklin** [comment]
This should work:

```python
dtypes_list = dask.compute([part.dtypes for part in df.to_delayed()])
```

File: dask/dataframe/tests/test_multi.py [hunk 1]
```diff
@@ -323,6 +323,23 @@ def test_concat(join):
...
+                         'y': list('abcdef')},
+                        index=[1, 2, 3, 4, 6, 7])
+    ddf1 = dd.from_pandas(pdf1, 2)
+    pdf2 = pd.DataFrame({'x': [1.0, 2.0, 3.0, 4.0, 6.0, 7.0],
+                         'y': list('abcdef')},
+                        index=[8, 9, 10, 11, 12, 13])
+    ddf2 = dd.from_pandas(pdf2, 2)
+
+    expected = pd.concat([pdf1, pdf2], join=join)
+    result = dd.concat([ddf1, ddf2], join=join)
```
Comment (mrocklin) [inline dask/dataframe/tests/test_multi.py]: You may also want to add the following

```suggestion
    result = dd.concat([ddf1, ddf2], join=join)
    assert_eq(expected, result)
```

**Comment by quasiben** [comment]
@glemaitre thanks for the PR.  Anything else you need help with ?

**Comment by glemaitre** [comment]
> @glemaitre thanks for the PR. Anything else you need help with ?

A bit of time :) We are realising scikit-learn so I might have some delay for finishing the PR.

**Comment by mrocklin** [comment]
I'll try to bang this out now.

**Comment by mrocklin** [comment]
Merging after passed tests

**Pull request merged**